src/_pytest/tmpdir.py

```diff
@@ -213,4 +266,46 @@ def tmp_path(request: FixtureRequest, tmp_path_factory: TempPathFactory) -> Path
     The returned object is a :class:`pathlib.Path` object.
     """
 
-    return _mk_tmp(request, tmp_path_factory)
+    path = _mk_tmp(request, tmp_path_factory)
+    yield path
+
+    # Remove the tmpdir if the policy is "failed" and the test passed.
+    tmp_path_factory: TempPathFactory = request.session.config._tmp_path_factory  # type: ignore
+    policy = tmp_path_factory._retention_policy
+    if policy == "failed" and request.node.result_call.passed:
+        rmtree(path)
+
+    # remove dead symlink
+    basetemp = tmp_path_factory._basetemp
+    if basetemp is None:
+        return
+    for left_dir in basetemp.iterdir():
```

Comment:
Lets move this block of code to a function in `pathlib.py` and reuse it in `cleanup_numbered_dir`. 👍

Reply:
Done with [b04bbb1](https://github.com/pytest-dev/pytest/pull/10442/commits/b04bbb1dcb91c327d747943862a53095c5d4269a).